### ISSUE TITLE
train_lm.py using LSTM results nan/inf

### ISSUE THREAD

**Issue by zhang-jian** [taolei87/sru#2]
Hi,
Did you use seed in your language model experiments? I tried to run LSTM experiment with default papameters, but the training was terminated at epoch 53. Looking at train_lm.py , I think this is because:
<pre>
if math.isnan(loss.data[0]) or math.isinf(loss.data[0]): 
</pre>

Thanks,

**Comment by taolei87** [comment]
@zhang-jian hi, no i didn't use an explicit seed in my experiment. the default option is for SRU, which uses 6 layers. i suspect LSTM gets gradient explosion. 

could you try `--depth 2` or `--depth 3` for LSTM? Plz let me know if still has the problem, and i'll take a look.

**Comment by chilynn** [comment]
Hi, I also meet the similar problem. I build my network which includes the sru layer. During the training phrase, the loss suddenly become very large and then become normal, sometimes even become nan. It seems to be the problem of gradient explosion. So I use the clip_grad_norm, but it dose not work. Finally, I replace the sru layer by lstm layer. The loss stably decrease. Could you please take a look. Thanks!

**Comment by taoleicn** [comment]
@chilynn I need a bit more details.
are u running the example `train_lm.py` I put in this repo, or your own code?

**Comment by chilynn** [comment]
I use my own code and data, below is my network which is very simple:

'''
self.sru = SRU(self.feature_dim,
	self.hidden_dim,
	num_layers = 1,          # number of stacking RNN layers
	dropout = 0.3,           # dropout applied between RNN layers
	rnn_dropout = 0.0,       # variational dropout applied on linear transformation
	use_tanh = 1,            # use tanh or identity activation
	bidirectional = False    # bidirectional RNN ?
)
	self.lstm = torch.nn.LSTM(self.feature_dim, self.hidden_dim)
	self.fc1 = torch.nn.Linear(self.hidden_dim, self.hidden_dim)
	self.fc2 = torch.nn.Linear(self.hidden_dim, self.output_dim)

output, hidden = self.sru(x)
output = output[-1, :, :]
output = self.fc1(output)
output = self.fc2(output)
'''

another network which only replace sru by lstm

'''
output, hidden = self.lstm(x)
'''

**Comment by chilynn** [comment]
@taoleicn I use my own code and data, here is my network which is very simple:

'''
def __init__(self, ...):
&nbsp; &nbsp; self.sru = SRU(self.feature_dim,
&nbsp; &nbsp; self.hidden_dim,
&nbsp; &nbsp; num_layers = 1,          # number of stacking RNN layers
&nbsp; &nbsp; dropout = 0.3,           # dropout applied between RNN layers
&nbsp; &nbsp; rnn_dropout = 0.0,       # variational dropout applied on linear transformation
&nbsp; &nbsp; use_tanh = 1,            # use tanh or identity activation
&nbsp; &nbsp; bidirectional = False    # bidirectional RNN ?
&nbsp; )
&nbsp; self.lstm = torch.nn.LSTM(self.feature_dim, self.hidden_dim)
&nbsp; self.fc1 = torch.nn.Linear(self.hidden_dim, self.hidden_dim)
&nbsp; self.fc2 = torch.nn.Linear(self.hidden_dim, self.output_dim)

def forward(self, x):
&nbsp; output, hidden = self.sru(x)
&nbsp; output = output[-1, :, :]
&nbsp; output = self.fc1(output)
&nbsp; output = self.fc2(output)
'''

another network which only replace sru by lstm

'''
output, hidden = self.lstm(x)
'''

**Comment by taolei87** [comment]
Based on my experience,

SRU tends to get larger gradient compared to LSTM.  As a result, SRU may need a stronger (smaller) gradient clipping compared to LSTM.

Also, not sure how you initialize the parameters. In my lm experiments, all values (including embeddings) are from [ -sqrt(3/d), sqrt(3/d) ], which works better:
https://github.com/taolei87/sru/blob/master/language_model/train_lm.py#L85-L91 <https://github.com/taolei87/sru/blob/master/language_model/train_lm.py#L85-L91>
